azkaban-common/src/test/java/azkaban/project/DirectoryYamlFlowLoaderTest.java

```diff
@@ -37,31 +52,136 @@ public void setUp() {
+    Assert.assertEquals(3, loader.getFlowMap().size());
+    Assert.assertTrue(loader.getFlowMap().containsKey(EMBEDDED_FLOW));
+    final Flow flow = loader.getFlowMap().get(EMBEDDED_FLOW);
+    Assert.assertEquals(4, flow.getNodes().size());
+
+    Assert.assertTrue(loader.getFlowMap().containsKey(EMBEDDED_FLOW_1));
+    final Flow flow1 = loader.getFlowMap().get(EMBEDDED_FLOW_1);
+    Assert.assertEquals(4, flow1.getNodes().size());
+
+    Assert.assertTrue(loader.getFlowMap().containsKey(EMBEDDED_FLOW_2));
+    final Flow flow2 = loader.getFlowMap().get(EMBEDDED_FLOW_2);
+    Assert.assertEquals(2, flow2.getNodes().size());
+
+    // Verify flow edges
+    Assert.assertEquals(3, loader.getEdgeMap().size());
+    Assert.assertEquals(3, loader.getEdgeMap().get(EMBEDDED_FLOW).size());
+    Assert.assertEquals(3, flow.getEdges().size());
+    for (final Edge edge : loader.getEdgeMap().get(EMBEDDED_FLOW)) {
+      this.logger
+          .info(EMBEDDED_FLOW + ".flow has edge: " + edge.getId());
+      Assert.assertNull(edge.getError());
+    }
+
+    Assert.assertEquals(3, loader.getEdgeMap().get(EMBEDDED_FLOW_1).size());
+    for (final Edge edge : loader.getEdgeMap().get(EMBEDDED_FLOW_1)) {
+      this.logger.info(
+          EMBEDDED_FLOW_1 + ".flow has edge: " + edge.getId());
+      Assert.assertNull(edge.getError());
+    }
+
+    Assert.assertEquals(1, loader.getEdgeMap().get(EMBEDDED_FLOW_2).size());
+    for (final Edge edge : loader.getEdgeMap().get(EMBEDDED_FLOW_2)) {
+      this.logger.info(EMBEDDED_FLOW_2 + ".flow has edge: " + edge.getId());
+      Assert.assertNull(edge.getError());
+    }
+  }
+
+  @Test
+  public void testLoadInvalidFlowYamlFiles() {
+    final DirectoryYamlFlowLoader loader = new DirectoryYamlFlowLoader(new Props());
+
+    loader.loadProjectFlow(this.project, ExecutionsTestUtil.getFlowDir(INVALID_FLOW_YAML_DIR));
+    Assert.assertEquals(2, loader.getErrors().size());
+    Assert.assertEquals(2, loader.getFlowMap().size());
+    Assert.assertEquals(2, loader.getEdgeMap().size());
+
+    // Invalid flow 1: Dependency not found.
+    Assert.assertTrue(loader.getFlowMap().containsKey(INVALID_FLOW_1));
+    final Flow flow1 = loader.getFlowMap().get(INVALID_FLOW_1);
+
+    Assert.assertEquals(3, flow1.getNodes().size());
+    Assert.assertEquals(1, flow1.getErrors().size());
+
+    Assert.assertEquals(3, loader.getEdgeMap().get(INVALID_FLOW_1).size());
+    Assert.assertEquals(3, flow1.getEdges().size());
+    for (final Edge edge : loader.getEdgeMap().get(INVALID_FLOW_1)) {
+      this.logger.info(INVALID_FLOW_1 + ".flow has edge: " + edge.getId());
+      if (edge.getError() != null) {
+        Assert.assertEquals(DEPENDENCY_NOT_FOUND_ERROR, edge.getError());
+      }
+    }
+
+    // Invalid flow 2: Cycles found.
+    Assert.assertTrue(loader.getFlowMap().containsKey(INVALID_FLOW_2));
+    final Flow flow2 = loader.getFlowMap().get(INVALID_FLOW_2);
+
+    Assert.assertEquals(4, flow2.getNodes().size());
```

Comment:
When parsing yaml runs into any issue, it looks like nodes are identified anyhow. What's the intention behind this behavior?

Reply:
Actually it happens in two steps, parsing the YAML file to AzkabanFlow first and then converting AzkabanFlow to Flow object. In this case, parsing the YAML file is successful, so we can identify the nodes. However, when converting AzkabanFlow to Flow, we resolve the dependency and detect an error. The error will be returned back in the report showing something wrong in the flow to users.